server/src/main/java/org/elasticsearch/index/mapper/BooleanFieldMapper.java

```diff
@@ -237,23 +277,53 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             return;
         }
 
-        Boolean value = context.parseExternalValue(Boolean.class);
-        if (value == null) {
-            XContentParser.Token token = context.parser().currentToken();
-            if (token == XContentParser.Token.VALUE_NULL) {
-                if (fieldType().nullValue() != null) {
-                    value = fieldType().nullValue();
+        Boolean value = null;
+        try {
+            value = context.parseExternalValue(Boolean.class);
+            if (value == null) {
+                final XContentParser parser = context.parser();
+                XContentParser.Token token = parser.currentToken();
+                if (token == XContentParser.Token.VALUE_NULL) {
+                    if (fieldType().nullValue() != null) {
+                        value = fieldType().nullValue();
+                    }
+                } else {
+                    if (indexCreatedVersion.onOrAfter(Version.V_6_0_0_alpha1)) {
+                        if(!parser.isBooleanValue()) {
+                            // need to consume array/object, otherwise parser doesnt skip to the next field as expected
+                            parser.skipChildren();
+
+                            throw new IllegalArgumentException(parser.currentToken().toString());
+                        }
+                        value = parser.booleanValue();
+                    } else {
+                        if (parser.isBooleanValueLenient()) {
+                            value = parser.booleanValueLenient();
+
+                            if (!parser.isBooleanValue()) {
+                                String rawValue = parser.text();
+                                deprecationLogger.deprecated("Expected a boolean for property [{}] but got [{}]",
```

Comment:
Could you explain why you log a deprecation her? Might be missing some context, but I thought this PR wasn't about deprecation but about adding some option to the field mapper.

Reply:
see my comment above about legacy lenient support recently removed.